### ISSUE TITLE
Matrix Survey Option

### ISSUE THREAD

**Issue by SuruPa00** [BIDMCDigitalPsychiatry/LAMP-platform#442]
Can we have a new survey option that looks like a matrix that can be created to reduce having multiple questions with the same instructions and options. 
![image](https://user-images.githubusercontent.com/86117379/144280098-6d525fb6-7b07-415c-bf30-0a22a683552b.png)

**Comment by sarithapillai8** [comment]
@SuruPa00 
This is updated in staging. Please check and let us know if this is ok to go production.

**Comment by SuruPa00** [comment]
Can you check on this? On the mobile, the options for the matrix are cutoff (I've allowed 5 options, but only 4 are visible) - can you fix this by making it so the bubbles/boxes are smaller and so at least 5 can fit on a screen and if they dynamically get smaller based on how many options there are so they can all fit on one screen? Additionally, there is no indicator above the bubbles showing what the option of each bubble indicates even though I have configured them to be 1-5. Can this be fixed so you can see that?

![Image from iOS (16)](https://user-images.githubusercontent.com/86117379/145829486-1d8c40b6-63b1-48fd-92d5-242928f1298a.png)

**Comment by sarithapillai8** [comment]
@SuruPa00 
This is updated in staging.

**Comment by SuruPa00** [comment]
@sarithapillai8 i've checked on this and we can update this and #447 to production once i get the logs

**Comment by sarithapillai8** [comment]
Please see the test report attached
[LAMP_Admin Website_Survey Matrix.xls](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/files/7716876/LAMP_Admin.Website_Survey.Matrix.xls)

**Comment by SuruPa00** [comment]
@sarithapillai8 this looks good, but the way it looks in the portal tab is not intuitive and does not make much sense. Would it be possible for the line graphs to show for each individual question added for the matrix? If there are 5 prompts within a matrix, there should be 5 lines.

**Comment by sarithapillai8** [comment]
This is updated in staging. Please check and let us know. QA is testing on the same.

**Comment by SuruPa00** [comment]
@sarithapillai8 I am not seeing any update in staging. This is the current single line graph i see even though there are various values for all 4 entries.

![image](https://user-images.githubusercontent.com/86117379/146384393-1c24ebbc-85bf-4140-ac7c-17c67aa8c5bb.png)

**Comment by sarithapillai8** [comment]
@SuruPa00 
 This is the graph for whole survey, right ? What we need to change there?
 I have changed the detail view.Can you check the detail view by clicking the icon on top left corner?

**Comment by SuruPa00** [comment]
@sarithapillai8 can we decrease the box dedicated to the circles for the matrix so that we have more room for texts? we need to be able to fit longer questions without making the size too small, so smaller amounts of space for the circles/boxes would be better

additionally, can you change question description color to be black instead of grey?

**Comment by sarithapillai8** [comment]
@SuruPa00 
The last change is updated in staging. Could you check and let me know if that is ok?

**Comment by SuruPa00** [comment]
@sarithapillai8 just checked, this looks good - i will wait to push to production as other items are being worked on by your team right now

**Comment by sarithapillai8** [comment]
@SuruPa00 
The only other item waiting for confirmation now is #400.Our QA confirmed that it is working fine and if you can check from your side, it is ready to go. No other changes will be doing on staging until then.

**Comment by divyav2020** [comment]
We have updated the build in Dashboard staging
[LAMP_Admin Website_Survey Matrix.xls](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/files/7829115/LAMP_Admin.Website_Survey.Matrix.xls)

**Comment by sarithapillai8** [comment]
@avaidyam 
This is done different question like slider, multiselect. The questions and answer options are saved as question options. In survey each question in different index and the whole matrix question is considered as single question. The Prevent graph is also handled with this kind of data. So can you confirm we need to change this implementation?

**Comment by avaidyam** [comment]
This implementation was deemed okay temporarily but let me get back to you on the more robust and correct implementation.

**Comment by lukeoftheshire** [comment]
Reproducing a conversation @avaidyam and I had regarding the best way to store this data:

 My gut instinct is that we should try to preserve as much data as possible, i.e. having some way to mark these as a matrix in the backend - I've always thought it would be interesting to, for example, look how (or if) different survey response options affect responses, so having the ability to reconstruct that would be nice.

But! From a duration/timestamp pov there needs to be some separation. Duration is currently not really usable as far as I can tell - grouping questions like this loses some data.

I would propose something like this (using the GAD-7 shown in the initial post as a base):
```
item:The specific question responded to (e.g. "Feeling nervous...")
value: response to specific question ("Not at all")
type: Matrix header ("over the last two weeks...")
level: (still none?)
duration: timestamp of THIS question 
```
I think this solves the following problems (which may be a little subjective):
- "item" tells us nothing currently, since "matrix question" can be inferred from the format of "value". It also doesn't match other survey question data formats where the "item" key stores the thing the participant saw/responded to.
- I dislike how basically all data is stored in 'value' tbh - I like nested dicts in general but this just feels like an overly complicated array since the keys are just 0-(n-1). (In fact it may be worse than an array since it can't be as easily iterated over)
- type and level are both unused so we might as well get something out of them
- by grouping multiple questions into one response duration becomes useless


I think we could also consider having "type" be "matrix", and then level (or another key if that's something we can do) be the matrix question. It can (in my experience) add some extra work to have to infer the question type from the value alone so adding some additional info might be nice.

This is of course just a proposal so I'm interested in what other people think.

**Comment by avaidyam** [comment]
One thing to consider is that the matrix header does not need to be recorded in the data output. I think the simplest way of doing this would be to make the `matrix` question type become an "accumulated" group that takes on the actual question type from the question before it. Then:

```
1. "How are you doing today?" (Boolean)
2. "In the past week, how often have you..." (Slider = 1-100)
3. "Brushed your teeth?" (Matrix)
4. "Washed your face?" (Matrix)
5. ...
6. "Another question" (Text)
```

Would render questions 2-5 as sliders 1-100 on the same page, since the earliest non-matrix question is q2. q6 would break that and be presented on the next page. Here, q3-5 have the expected data output of a slider question type, and q2 would have `value = null`. This is also backwards compatible with non-matrix questions, sort of. Thoughts, @lukeoftheshire?

**Comment by lukeoftheshire** [comment]
There are a couple concerns I have with this implementation (although please correct me if I misunderstood anything).

First - I'm not sure about the structure of the data. I think with a "one slice is one question/event" approach it is easy from a data analysis perspective to iterate over an array of events and handle each one. With something like this it could be easy to fall into a pit where you keep having to refer back to the last time you had a 'non-matrix' type to figure out the question asked. It would be doable but perhaps messy.

Second - I think using `value = null` is a dangerous way to identify the matrix "header" alone - we just implemented optional questions, for instance. I assume the value for those is `null` if left unanswered? If so, we run the risk of falsely identifying something as a matrix. Or, we may want to save 'value=null' for future questions, in which case it wouldn't uniquely identify matrix questions.

If we *do* decide to do  something like this I think we should also consider the reverse, where the "in the past week" is set to indicate "matrix" somehow, and then each question after is set to "Slider = 1-100". This would largely solve the two issues above, I think. However, this raises the issue of figuring out when the matrix ends, so this is perhaps not a perfect solution.

**Comment by avaidyam** [comment]
> we just implemented optional questions, for instance

That's a great point - it breaks consistency with our API usage and is confusing to end-users of the API or Cortex. 

> then each question after is set to "Slider = 1-100"

I'm not following this part - do you have an example JSON for this so I can visualize it?

**Comment by lukeoftheshire** [comment]
Maybe something like this:
```
[{'item': 'In the past week, how often have you?',
     'value': None,
     'type': 'matrix',
     'level': None,
     'duration': 777},
    {'item': 'Test Slider',
     'value': 100,
     'type': 'slider',
     'level': None,
     'duration': 4871},
    {'item': 'Test Slider 2',
     'value': 0,
     'type': 'slider',
     'level': None,
     'duration': 777}]
```

Basically, I think we need to *strongly* consider leveraging the unused 'type' variable. We should probably be considering this anyway - from our normal data structure it is basically impossible to figure out what question is what without using the stored activity settings - which is not ideal considering those can be changed. 

However the downside to this is it is still unclear where a matrix ends. Because of that I think including the matrix header in each temporal slice somehow (maybe in level?) might be a good idea. Or we could change type to `slider-matrix` which might also help. This is more me throwing out ideas.

**Comment by avaidyam** [comment]
> it is still unclear where a matrix ends

One point to note is that a matrix must have every question be of the same type. 

> change type to slider-matrix which might also help

I am somewhat opposed to this - we would duplicate most question types and make them more difficult to parse this way.

**Comment by lukeoftheshire** [comment]
>One point to note is that a matrix must have every question be of the same type.

That's a good point - if we implement type tracking the only situation we might get confused is if, for example, a non-matrix slider follows a matrix of sliders.

I agree that 'slider-matrix' is a klunky implementation! Definitely not committed to it, more suggesting things to see if anything sticks. That being said I still think we should consider if there is any way to have the matrix info contained in each slice - I am very much in favor of each temporal slice allowing us to fully reconstruct what was shown to the participant without needing to fetch data from other resources like the Activity configs or even elsewhere in the temporal slice array.

**Comment by avaidyam** [comment]
> each temporal slice allowing us to fully reconstruct what was shown to the participant without needing to fetch data from other resources

Perhaps this is a different issue to track? For example, the API could potentially return the set of Activity objects referenced by each group of requested ActivityEvents. If you think that's an approach that might work but needs some discussion, let's make a new issue for that!

**Comment by avaidyam** [comment]
Currently, the `matrix` question UI offers a list-builder with: 
1. `Enable multiple option selections`
    - This is presumably to switch the matrix questions from `list` to `multiselect`. 
    - **We're reimplementing the wheel here.**
1. Questions
    - This section is obviously to add the nested questions.
    - **It doesn't currently support question descriptions, which means we're now deviating from our norm which can and will confuse users.**
1. Options
    - This section adds the common list choices for all the nested questions. 
    - **This implies all of the questions have the same exact scale/list - which is good because it enforces the "matrix" style.** 

---

Looking at this now, I think `matrix` questions should be simplified as per my original suggestion. 
1. To create a matrix question, create your first question like so:
    1. The question title will become the **title of the first matrix question**: `"Brushed your teeth"`
    2. The question description will become the **title of the matrix section**: `"In the past week, how many times have you..."`
    3. The question type will become the **type of ALL the matrix questions**: `list`/`multiselect` make sense here, but `text` will also work exactly the same way[^1].
        1. The configuration of the question options (where available, since `text` doesn't have any configuration, etc.) will become the **options/configuration of ALL the matrix questions**: `["Never", "Rarely", "Often", "Always"]`
        2. The `optional` toggle will apply only to this first question in the matrix section.
2. To add questions to this matrix (notice how the first question is just a normal question otherwise), create `n` following questions like so: 
    1. The question title will become the **title of the n-th matrix question**: `"Washed your face"`, etc.
    2. **The question description will be ignored entirely for matrix questions** _(Since it is only a visual element and not related to the data output of the survey.)_
    3. The question type **MUST be set to `matrix`.**
        1. **There are no configuration/options** for the `matrix` question type. (If absolutely required in the future, for example, we could add support for additional parameters.)
        2. The `optional` toggle will apply to the `n-th` question in the matrix section. (That is, the "optionability" of each question in the matrix is independent of other questions.)
3. To end the matrix question section and switch to a different question type for succeeding questions, simply create a question that is **NOT** a `matrix` type. 
    - If a `matrix` type question is the last one in the survey, it's the same as ending the matrix section. 
    - If a `matrix` type question is the **FIRST** one in the survey (which is technically illegal configuration), it should be treated as a `text` question, and subsequent matrix questions must assume the same fact.

---

This configuration setup will ensure the following: 
1. `ActivityEvent` output for a matrix section of 10 questions **WILL and MUST** look identical to those same 10 questions configured identically. 
2. There **WILL NOT** be a blank/placeholder question item in the `Activity.settings` OR in `ActivityEvent.temporal_slices`.
3. All data visualization/output processing may assume matrix questions do not exist, logically. 
     - When looking up `Activity.settings`, if the question has type `matrix`, the client must walk-back to the last-known non-`matrix` type. 
     - If there is no last-known non-`matrix` type (i.e. the first question is illegally set as `matrix`), all matrix questions in that section should be treated as `text` question types.

[^1]: Matrix questions are just a different visual arrangement of the same questions we already support. They should not offer additional functionality OR change any data/configuration & input/output.

**Comment by avaidyam** [comment]
@sarithapillai8 Can you review the above comment **in extreme depth** and let me know if you have any questions about the implementation strategy? //cc @lukeoftheshire

**Comment by sarithapillai8** [comment]
@avaidyam 
Please see my understanding. Correct me if anything listed wrong.

1. Start matrix question when a question is created with question type **_list/multiselect/text_** and following questions are with question type _**matrix**_ (_no configuration except 'optional' will be available for matrix question. So it will look similar to text type_).
2.  The question title, description can be considered as the matrix section title and description.
3. The options used for matrix questions can be the options provided for list/multiselect question if the first question created as list/multiselect, and If the first question is text type, use ["Never", "Rarely", "Often", "Always"] as options.
4. End of matrix can be identified with a question with any question type other than matrix or when there are no more questions in the survey.
5.  Each matrix question will have an separate entry in temporal_slices of ActivityEvent.
6.  When a question is created with question type list/multiselect/text, and following questions are not matrix type, then treat  them as normal list/multiselect/text question. 
7. Also when a question is created as matrix question without any preceding list/multiselect/text question, 
treat them as text type question.

>The optional toggle will apply to the n-th question in the matrix section. (That is, the "optionability" of each question in the matrix is independent of other questions.)

So this is like the optional selection(currently the text is changed to mandatory with Survey JSON Schema change) will be available to all matrix questions like other questions, thus the _optional checking required for each matrix question_.

**Comment by avaidyam** [comment]
> Start a matrix question when a question is created with question type **list/multiselect/text**

🚫 The starting question type can be **anything** (slider, rating, short answer, etc. too) **INCLUDING** `matrix`. If the first question in the survey is of type `matrix`, that's an illogical edge-case, so treat it as if that first question was actually `text` instead. 

> following questions are with question type matrix

✅ 

> no configuration except 'optional' will be available for matrix question. So it will look similar to text type

✅ _(By "look similar" I assume you mean the editor/configuration side, which is correct. If not, please comment what you mean.)_

> The question title, description can be considered as the matrix section title and description.

⚠️ **Yes, but in reverse order.** The title stays as the question text, but the description becomes the whole matrix section's title. **No question descriptions will be displayed for a matrix section.** 
```
The question title will become the title of the first matrix question. The question description will become the title of the matrix section.
```

> The options used for matrix questions can be the options provided for list/multiselect question if the first question created as list/multiselect

⚠️ **Yes, but this applies to ALL question types,** not just `list` or `multiselect`. If the first question is a `slider` type with range `0 - 100`, that slider configuration applies to all `matrix` questions in the section too. 

> If the first question is text type, use ["Never", "Rarely", "Often", "Always"] as options.

🚫 This was not stated. `["Never", "Rarely", "Often", "Always"]` were provided as **EXAMPLES** of what a clinician would use for a `list` matrix. If the first question is a `text` type, all questions in the matrix section should be text boxes.

> End of matrix can be identified with either a question with any question type other than matrix or when there are no more questions in the survey.

✅ 

> Each matrix question will have a separate entry in temporal_slices of ActivityEvent.

✅ 

> But what to do with the first question - list/multiselect/text type which one considered to be the title part of matrix section, can this be ignored for temporal_slices entry?

🚫 No, this question **IS the first question in the matrix section.** For example, `settings = [q0, ..., q5[list], q6[matrix], q7[matrix], q8[matrix], q9[slider], ..., q15]` will produce `temporal_slices = [a0, ..., a5, a6, a7, a8, a9, ...., a15]`. (Each question has an exactly matching answer.)

> When a question is created with question type list/multiselect/text, and following questions are not matrix type, then treat them as normal list/multiselect/text question.

✅ _(Yes, but as stated above, this is true for ANY non-`matrix` question type, not just `list`/`multiselect`/`text`.)_

> Also when a question is created as matrix question without any preceding list/multiselect/text question, treat them as text type question.

⚠️ If a survey consists of **FIVE** questions, **ALL** of which are `matrix` type, then **ONLY THE FIRST question** should be treated as a `text` type. All other questions are still `matrix` type. (i.e. convert `q1-5 = matrix` to `q1 = text, q2-5 = matrix` at runtime when the survey UI is opened by the patient.)

> In this case the question description will be shown on the survey, otherwise no use. 

🚫 No, question descriptions will NOT be shown for any questions part of a matrix section. In the case above, where we convert `q1 = matrix` to `q1 = text`, the question description, if provided, will be used as the matrix section title, as if `q1` was always configured as `text`.

> Or as you said, do we need to hide question description for matrix type question? it comes above the question type and common for all questions. 

⚠️ Except for this `q1 = matrix` edge case, question descriptions are hidden for all questions in the matrix section when the survey is opened by a patient. 

> Not sure if this can be handled in JSON schema easily, will check and let you know if that required.

⚠️ The JSON Schema side should stay the same - the `question description` field will still be present, even for `matrix` type questions, and clinicians can fill them out as they currently do, but they will just be ignored by the Survey UI when opened by the patient (and thus not rendered on screen). 

> So this is like the optional selection will be available to all matrix questions like other questions, thus the optional checking required for each matrix question.

✅ 

> (currently the text is changed to mandatory with Survey JSON Schema change)

⚠️ Isn't the `Activity.settings` field called `optional` currently? The JSON Schema Form should reflect that and keep the field title `Optional`. If the field is called `required`, then the Form field title should also be called `Required`.

**Comment by sarithapillai8** [comment]
> ✅ (By "look similar" I assume you mean the editor/configuration side, which is correct. If not, please comment what you mean.)

Yes I meant that.
>The starting question type can be anything (slider, rating, short answer, etc. too) INCLUDING matrix. If the first question in the survey is of type matrix, that's an illogical edge-case, so treat it as if that first question was actually text instead.

Got it

>This was not stated. ["Never", "Rarely", "Often", "Always"] were provided as EXAMPLES of what a clinician would use for a list matrix. If the first question is a text type, all questions in the matrix section should be text boxes.

I thought it is like the default options like - ["Nearly All the Time", "More than half the time", "Several time", "Not at all].  So if the first question is of type text/short/time, show the following matrix questions like text questions in grid, or just ignore matrix type and show them as normal text type questions?

>No, this question IS the first question in the matrix section. For example, settings = [q0, ..., q5[list], q6[matrix], q7[matrix], q8[matrix], q9[slider], ..., q15] will produce temporal_slices = [a0, ..., a5, a6, a7, a8, a9, ...., a15]. (Each question has an exactly matching answer.)

Got it.

>Except for this q1 = matrix edge case, question descriptions are hidden for all questions in the matrix section when the survey is opened by a patient.

So first question description will be the section title of the matrix , and title will be the first question. All the following matrix question descriptions will have no role in UI.

>Isn't the Activity.settings field called optional currently? The JSON Schema Form should reflect that and keep the field title Optional. If the field is called required, then the Form field title should also be called Required.

Activity.settings field is 'required'.  Form field now used is Mandatory. We can update it to required.

**Comment by avaidyam** [comment]
> So if the first question is of type text/short/time, show the following matrix questions like text questions in grid, or just ignore matrix type and show them as normal text type questions?

You should show them as text/short/time questions in a grid layout. 👍 

_(It might look weird and not optimized for a mobile layout, but that is what some collaborators seem to want, so my team will share screenshots of good and bad examples on a phone screen layout to explain that it's not helpful. This should also simplify the code for matrix section layouts.)_

> So first question description will be the section title of the matrix , and title will be the first question. All the following matrix question descriptions will have no role in UI.

Exactly. 👍 

> Activity.settings field is 'required'. Form field now used is same and label is Mandatory. I can change the label too.

Got it, please use `Required`. 👍

**Comment by divyav2020** [comment]
[LAMP_Admin Website_Survey Matrix New changes.xls](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/files/8189960/LAMP_Admin.Website_Survey.Matrix.New.changes.xls)

**Comment by sarithapillai8** [comment]
@avaidyam 
Sorry for the late clarification. If the first question is a slider type question, and following are matrix questions, currently it is showing with radio options like attached.
![2CBCB0E5-B65C-49E5-B348-B1742A091DA3](https://user-images.githubusercontent.com/63710347/157396345-61c93e0f-cc56-4ea8-acca-c584b753607f.jpg)

Do you actually want to show the slider component in matrix format in this case?

**Comment by avaidyam** [comment]
Yes, it needs to show the actual `slider` or `rating` component instead of the `list` (radio button) component.

**Comment by lukeoftheshire** [comment]
A matrix of short answers does not display the text box:
<img width="789" alt="Screen Shot 2022-04-12 at 1 45 23 PM" src="https://user-images.githubusercontent.com/34659264/163023864-81b19175-26d3-40fd-a4ce-e5b81b6835e7.png">

If one or more of the questions are required the survey cannot be completed.
<img width="1121" alt="Screen Shot 2022-04-12 at 1 46 14 PM" src="https://user-images.githubusercontent.com/34659264/163023766-23292812-6b24-405f-88b3-5cb60c9bb557.png">

**Comment by sarithapillai8** [comment]
@lukeoftheshire 
This is updated in dashboard-staging. The short answer question had an options array, which might be because of editing it from a list/multiselect/slider/rating question type. So the fix done for that section. Otherwise short answer is working fine.
![image](https://user-images.githubusercontent.com/63710347/163107129-6c0b441a-21eb-4667-877a-d3b0badb2cf1.png)

>If one or more of the questions are required the survey cannot be completed.

This is an expected behavior I hope. As the short answer were missing input fields, and they are required questions - the survey could not be submitted.

**Comment by lukeoftheshire** [comment]
QA tested this, I can confirm the short answers now work! And yes - being unable to complete the survey is expected behavior - I was just mentioning it to underscore that we needed to fix this prior to release.

One other thing I noticed when testing is that it looks like the format of survey temporal slices has changed. For instance, here you can see a slice from a new survey:

```
    {'item': 'Matrix Short Answer 2',
     'value': 'Asdf',
     'duration': 422,
     'endTime': 1649960650824},
```
where an old slice might look like:
```
{'item': 'Overall, how would I rate my anxiety today ?',
     'value': '3',
     'type': None,
     'level': None,
     'duration': 13480}
```

Type and level were unused before so I don't think this is a big issue, but I think this would take us away a bit from the "standard" temporal slice formatting and adds in endTime. Is this an intended change?

**Comment by avaidyam** [comment]
This should not be the case. `endTime` is not a supported parameter at all, and `type` still needs to be provided. @sarithapillai8 Can you resolve this?

**Comment by ZCOEngineer** [comment]
@avaidyam is there anything pending on this issue?

**Comment by sarithapillai8** [comment]
@avaidyam 
Yes. It was updated.

**Comment by Devayani0102** [comment]
This seems to fixed in staging
